Paternity Test.py

```diff
@@ -0,0 +1,378 @@
+#
+# | v | date | author | notes |
+# | --- | --- | --- | --- |
+# | 0.9 | April 6th 2020 | Ash | initial |
+# | 1.0 | April 13th 2020 | Ash, Lon |   fixed multiple children and no parent usecases, implented Lon's no-pandas code, implemented Brian and Lon's other suggestions, better intro|
+# | 1.1 | April 14th 2020 | Ash |   made Pandas code significantly faster|
+# | 1.2 | April 27th 2020 | Ash |   fixed NameError exception notice, rearranged for clarity and consistency across notebooks|
+#
+
+# # Setup
+
+# ## Imports
+
+# In[ ]:
+
+
+import io
+import os
+import firecloud.api as fapi
+import google.cloud.storage
+
+# Only used in multiple children use case
+import pandas as pd
+import logging
+
+
+# ## User-set variables
+
+# In[ ]:
+
+
+# Don't forgot the quotation marks!
+SUBDIRECTORY="/thousands/"
+TABLE_NAME="paternity_test_test" #Do not include spaces or weird characters
+PARENT_FILETYPE="cram"
+
+# If your filenames are like this, please set INCLUDE_PARENT_EXTENSION to True:
+# NWD119844.CRAM
+# NWD119844.CRAM.CRAI
+
+# If you filenames are like this, please set INCLUDE_PARENT_EXTENSION to False:
+# NWD119844.CRAM
+# NWD119844.CRAI
+
+# No quotation marks for this variable, just True or False
+INCLUDE_PARENT_EXTENSION = True
+
+
+# Make sure to include slashes in your SUBDIRECTORY variable.
+
+# ## Environmental variables
+
+# In[ ]:
+
+
+BILLING_PROJECT_ID = os.environ['GOOGLE_PROJECT']
+WORKSPACE = os.path.basename(os.path.dirname(os.getcwd()))
+BUCKET = os.environ["WORKSPACE_BUCKET"]
+
+
+# ## Call FireCloud
+
+# In[ ]:
+
+
+try:
+    response = fapi.list_entity_types(BILLING_PROJECT_ID, WORKSPACE)
+    if response.status_code != 200:
+        print("Error in Firecloud, check your billing project ID and the name of your workspace.")
+        raise
+    else:
+        print("Firecloud has found your workspace!")
+        directory = BUCKET + SUBDIRECTORY
+except NameError:
+    print("Caught a NameError exception. This may mean the kernal was restarted or you didn't run ",
+          "the cells above. Try running the cells above again.")
+    raise
+
+
+# ## Display the contents of your workspace bucket
+# This step is *optional* and you might want to skip it if you have a lot of files in your bucket.
+
+# In[ ]:
+
+
+get_ipython().system('gsutil ls $directory')
+
+
+# If you get a CommandError exception here, it may be because your SUBPREFIX is for a psuedo-folder that doesn't actually exist. Try `!gsutil ls $bucket` and make sure the directory you're looking for actally exists. If it's not there, run the Folder-Maker notebook first.
+
+# ## Delete placeholder file (if you created folder with Folder Maker)
+# Since there's now other files in the psuedo-folder (see Folder Maker for more info on why we are calling this a psuedo-folder), you can delete the placeholder file that Folder Maker made in order to prevent it from showing up in your TSV.
+
+# In[ ]:
+
+
+get_ipython().system('gsutil rm gs${directory}placeholder')
```

Comment:
I disagree; if the placeholder file is not deleted, then it will appear in the dataframe. Even worse, if it runs in my code, the placeholder will itself be considered a child file and will vastly slowdown the process as it tries in vain to find a parent file for it. At the same time I don't want my code to exclude "placeholder" as a file in case the user uploads a file with that name. How about I just tell the user to expect a "No URLs matched" CommandException if they already deleted the placeholder file?

Regarding the weird syntax: I don't know _why_ because directory as a variable definitely has gs in it, if I delete the gs in front of directory, I get an error. It's driving me nuts.

```
!gsutil rm ${directory}placeholder
InvalidUrlError: Unrecognized scheme "".

```
I can confirm that for reasons I don't understand, it does work correctly as written.